File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Analyzers.cs
using System.IO;
using System.Linq;
using Raven.Client.Documents;
using Raven.Client.Documents.Indexes;
using Raven.Client.Documents.Operations.Indexes;
using Raven.Documentation.Samples.Orders;


namespace Raven.Documentation.Samples.Indexes
{
    public class Analyzers
    {
        private class SnowballAnalyzer
        {
        }

        #region analyzers_1
        public class BlogPosts_ByTagsAndContent : AbstractIndexCreationTask<BlogPost>
        {
            public BlogPosts_ByTagsAndContent()
            {
                Map = posts => from post in posts
                               select new
                               {
                                   post.Tags,
                                   post.Content
                               };

                Analyzers.Add(x => x.Tags, "SimpleAnalyzer");
                Analyzers.Add(x => x.Content, typeof(SnowballAnalyzer).AssemblyQualifiedName);
            }
        }
        #endregion

        #region analyzers_3
        public class Employees_ByFirstAndLastName : AbstractIndexCreationTask<Employee>
        {
            public Employees_ByFirstAndLastName()
            {
                Map = employees => from employee in employees
                                   select new
                                   {
                                       LastName = employee.LastName,
                                       FirstName = employee.FirstName
                                   };

                Indexes.Add(x => x.FirstName, FieldIndexing.Exact);
            }
        }
        #endregion

        #region analyzers_4
        public class BlogPosts_ByContent : AbstractIndexCreationTask<BlogPost>
        {
            public BlogPosts_ByContent()
            {
                Map = posts => from post in posts
                               select new
                               {
                                   Title = post.Title,
                                   Content = post.Content
                               };

                Indexes.Add(x => x.Content, FieldIndexing.Search);
            }
        }
        #endregion

        #region analyzers_5
        public class BlogPosts_ByTitle : AbstractIndexCreationTask<BlogPost>
        {
            public BlogPosts_ByTitle()
            {
                Map = posts => from post in posts
                               select new
                               {
                                   Title = post.Title,
                                   Content = post.Content
                               };

                Indexes.Add(x => x.Content, FieldIndexing.No);
                Stores.Add(x => x.Content, FieldStorage.Yes);
            }
        }
        #endregion

        public Analyzers()
        {
            using (var store = new DocumentStore())
            {
                #region analyzers_2
                store.Maintenance.Send(new PutIndexesOperation(new IndexDefinitionBuilder<BlogPost>("BlogPosts/ByTagsAndContent")
                {
                    Map = posts => from post in posts
                                   select new
                                   {
                                       post.Tags,
                                       post.Content
                                   },
                    Analyzers =
                    {
                        {x => x.Tags, "SimpleAnalyzer"},
                        {x => x.Content, typeof(SnowballAnalyzer).AssemblyQualifiedName}
                    }
                }.ToIndexDefinition(store.Conventions)));
                #endregion
            }
        }
    }

    /*
    #region analyzers_6
    public class MyAnalyzer : Lucene.Net.Analysis.Analyzer
    {
        public override TokenStream TokenStream(string fieldName, TextReader reader)
        {
            throw new CodeOmitted();
        }
    }
    #endregion
    */
}

File: Documentation/5.2/Samples/nodejs/indexes/analyzers.js
import {
    AbstractIndexCreationTask,
    DocumentStore,
    IndexDefinition,
    PutIndexesOperation,
    IndexDefinitionBuilder, PutAnalyzersOperation, PutServerWideAnalyzersOperation
} from "ravendb";

const store = new DocumentStore();
const session = store.openSession();

class SnowballAnalyzer { }

//region analyzers_1
class BlogPosts_ByTagsAndContent extends AbstractIndexCreationTask {
    constructor() {
        super();

        this.map = `docs.Posts.Select(post => new {     
            tags = post.tags,     
            content = post.content 
        })`;

        this.analyze("tags", "SimpleAnalyzer");
        this.analyze("content", "Raven.Sample.SnowballAnalyzer");
    }
}
//endregion

//region analyzers_3
class Employees_ByFirstAndLastName extends AbstractIndexCreationTask {
    constructor() {
        super();

        this.map = "docs.Employees.Select(employee => new { " +
            "    LastName = employee.LastName, " +
            "    FirstName = employee.FirstName " +
            "})";

        this.index("FirstName", "Exact");
    }
}
//endregion

//region analyzers_4
class BlogPosts_ByContent extends AbstractIndexCreationTask {
    constructor() {
        super();

        this.map = "docs.Posts.Select(post => new { " +
            "    tags = post.tags, " +
            "    content = post.content " +
            "})";

        this.index("content", "Search");
    }
}
//endregion

//region analyzers_5
class BlogPosts_ByTitle extends AbstractIndexCreationTask {
    constructor() {
        super();

        this.map = "docs.Posts.Select(post => new { " +
            "    tags = post.tags, " +
            "    content = post.content " +
            "})";

        this.index("content", "No");
        this.store("content", "Yes");
    }
}
//endregion

async function analyzers() {
    const session = store.openSession();

    //region analyzers_2
    const builder = new IndexDefinitionBuilder("BlogPosts/ByTagsAndContent");
    builder.map = `docs.Posts.Select(post => new {     
        tags = post.tags,     
        content = post.content 
    })`;
    builder.analyzersStrings["tags"] = "SimpleAnalyzer";
    builder.analyzersStrings["content"] = "Raven.Sample.SnowballAnalyzer";

    await store.maintenance
        .send(new PutIndexesOperation(
            builder.toIndexDefinition(store.conventions)));
    //endregion

}

{
    //region analyzer_definition
    const analyzerDefinition = {
        name: "analyzerName",
        code: "code"
    };
    //endregion
}

let analyzerDefinition;
{
    //region put_analyzers_operation
    await store.maintenance.send(new PutAnalyzersOperation(analyzerDefinition));
    //endregion
}
{
    //region put_serverWide_analyzers_operation
    await store.maintenance.send(new PutServerWideAnalyzersOperation(analyzerDefinition));
    //endregion
}
{
    //region analyzers_7
    const analyzerDefinition = {
        name: "MyAnalyzer",
        code: "using System.IO;\n" +
            "using Lucene.Net.Analysis;\n" +
            "using Lucene.Net.Analysis.Standard;\n" +
            "\n" +
            "namespace MyAnalyzer\n" +
            "{\n" +
            "   public class MyAnalyzer : Lucene.Net.Analysis.Analyzer\n" +
            "    {\n" +
            "        public override TokenStream TokenStream(string fieldName, TextReader reader)\n" +
            "        {\n" +
            "            throw new CodeOmitted();\n" +
            "        }\n" +
            "    }\n" +
            "}\n"
    };

    await store.maintenance.send(new PutAnalyzersOperation(analyzerDefinition))
    //endregion
}



